Add arguments for sendmmsg and recvmmsg

Molter73 opened the pull request:
This is a basic implementation for the recvmmsg and sendmmsg syscalls in the CORE BPF probe. 

In order to simplify the implementation and assuming these syscalls are equivalent to multiple calls to recvmsg and sendmsg respectively, handling is done by mapping one call of recvmmsg/sendmmsg to multiple recvmsg/sendmsg events. This is a very naive implementation and is probably not correct enough to be upstreamed, but it is enough to get the events we need in order to support UDP on OCP Virt.

Molter73 committed:
Basic implementation for sendmmsg and recvmmsg

sbrivio-rh commented on driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/recvmmsg.bpf.c:
```diff
@@ -43,21 +50,72 @@ int BPF_PROG(recvmmsg_x,
+	struct mmsghdr *mmsghdr_pointer = (struct mmsghdr *)args[1];
 
-	// Here we have no parameters to collect.
+	for (int i = 0; i < ret && i < 16; i++)
```
The bound on `i` should be `UIO_MAXIOV` (1024) instead of 16, otherwise one can receive more than 16 messages and bypass the tracing for messages with indices greater than 15.

Molter73 replied:
Yeah, this is capped due to a verifier limitation, as it is now, it is being rejected on RHEL 8 kernels and if I bump it to 32 I think it gets rejected universally, so I'll need to change this to a `bpf_tailcall` approach or something more universal.

sbrivio-rh commented on driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/sendmmsg.bpf.c:
```diff
@@ -15,21 +16,66 @@ int BPF_PROG(sendmmsg_e,
-	/*=============================== COLLECT PARAMETERS  ===========================*/
-
-	ringbuf__submit_event(&ringbuf);
+	for (int i = 0; i < vlen && i < 16; i++)
```
Same here about the bound on `i`.

sbrivio-rh commented on driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/sendmmsg.bpf.c:
```diff
@@ -43,21 +89,60 @@ int BPF_PROG(sendmmsg_x,
+	struct mmsghdr *mmsghdr_pointer = (struct mmsghdr *)args[1];
 
-	// Here we have no parameters to collect.
+	for (int i = 0; i < ret && i < 16; i++)
```
Same here.

sbrivio-rh reviewed (commented):
Just one detail (repeated three times).

Molter73 committed:
Compatibility changes after Falco version bump

Molter73 committed:
Prevent truncation of verifier log

Molter73 committed:
Tweaks to modern_bpf implementation of sendmmsg and recvmmsg

Molter73 committed:
Add try_submit_event

Molter73 committed:
Minor tweaks

Molter73 committed:
Fixes to recvmmsg

Molter73 committed:
Cleanups

Molter73 committed:
Minor adjustments after upstream PR was opened

Molter73 committed:
Revert some unneeded changes

Molter73 committed:
Collect all sendmmsg parameters in the exit event

Molter73 committed:
Cleanups

Molter73 committed:
Fix verifier issue on Power and Z

These 2 architectures don't allow to tail call when bpf-to-bpf calls are
involved in non-JITted programs, so we prevent that from happening and
tail call from the main program instead.

Molter73 committed:
Remove hotplug tail calls

bpf_loop forces the called function to be a bpf-to-bpf call. In older
kernels, tail calls are not allowed when using bpf-to-bpf calls, so we
just get rid of the hotplug tail call, since we don't really use it
AFAIK.

Molter73 committed:
Remove some final tail calls

Molter73 committed:
Minor cleanups

Molter73 commented on driver/modern_bpf/helpers/store/auxmap_store_params.h:
```diff
@@ -153,7 +145,50 @@ static __always_inline void auxmap__submit_event(struct auxiliary_map *auxmap, v
+ * @param auxmap pointer to the auxmap in which we have already written the entire event.
+ * @param ctx BPF prog context
+ */
+static __always_inline void auxmap__submit_event(struct auxiliary_map *auxmap, void* ctx)
```
Just as a note on why this function had to be split into a 3, a base, a try and this one:

The new implementation for sendmmsg/recvmmsg using bpf_loop require the function passed as a callback parameter to be just that, a callback, meaning clang is not able to inline it like it does with all other functions. This in turn means that these new tracepoints produce bpf-to-bpf calls which are only allowed to do tail calls in JITed programs. That's all well and good, it works perfectly on x86, but it turns out it was implemented on arm, s390x and ppc64le only a couple years ago, which means it breaks a bunch of our integration tests with a verifier error. In order to fix this, I split the `auxmap__submit_event` so I could create a separate version that does not do the tail call, thus having our bpf-to-bpf call programs not do any tail calling themselves.

For additional context, the tail call is responsible for triggering a restart when a hotplug CPU is brought online, this is obviously crucial and should be done ASAP, not doing that tail call in these syscalls means that we will delay restart until another syscall is triggered. Given that restarting collector takes a non trivial amount of time, I think it's fine that we miss a couple syscalls, since the restart will cause us to drop quite a bit more than those.

Molter73 commented:
In order to work around a verifier issue on somewhat older kernels, #84 will make it so uninteresting sycalls will not be loaded into the kernel. That PR will be merged into this one, but in order to prevent bloat and make it easier to review the changes, they are separate for now.

If you are reviewing this PR, please take a minute to look into that one as well.

Stringy commented on driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/recvmmsg.bpf.c:
```diff
@@ -38,28 +44,131 @@ int BPF_PROG(recvmmsg_e,
+		return 0;
+	}
+
+	for(int i = 0; i < ret && i < 12; i++)
```
I take it 12 is the upper limit for an unrollable loop in BPF before the introduction of `bpf_loop`? Could we have a comment here to that effect?

Molter73 replied:
Yeah, this was trial and error against a verifier until it let me load the probe. Since I did the test some of the logic has changed, the error used to be due to the limit of processed instructions, but last time I tried it failed because of complexity since `handle_exit` is not being inlined. I'll try again on an older kernel and try to figure out if 12 is still the limit and add a comment.

Stringy commented on driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/sendmmsg.bpf.c:
```diff
@@ -25,7 +24,7 @@ int BPF_PROG(sendmmsg_e,
 	/*=============================== COLLECT PARAMETERS  ===========================*/
 
-	// Here we have no parameters to collect.
+	// Here we have no parameter to collect.
```
[nit] I'd argue the plural is correct in this case

Stringy commented on driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/sendmmsg.bpf.c:
```diff
@@ -38,26 +37,131 @@ int BPF_PROG(sendmmsg_e,
+	 * the return value if the syscall is successful.
+	 */
+	uint16_t snaplen = maps__get_snaplen();
+	// TODO: check sockaddr
```
is this still required?

Molter73 replied:
I need to look into it, I think we can pass in sockaddr as the last parameter to try and extract ip/port from it, but most of the time this information is provided by the kernel just fine.

Molter73 replied:
I'm an idiot, this only applies for upstream, we don't do dynamic snaplen and I think the function has actually changed its signature, I'll remove the comment.

Stringy commented on driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/sendmmsg.bpf.c:
```diff
@@ -38,26 +37,131 @@ int BPF_PROG(sendmmsg_e,
+	auxmap__store_iovec_data_param(auxmap, (unsigned long)mmh.msg_hdr.msg_iov, mmh.msg_hdr.msg_iovlen, snaplen);
+
+	/* Parameter 5: tuple (type: PT_SOCKTUPLE)*/
+	/* TODO: Check if the fd is a socket */
```
As above, do we need this?

Molter73 replied:
It is, I just don't really know how to test for it, the comment comes from another comment somewhere else in Falco's codebase. The way it's handled now should be fine though.

Stringy commented on driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/sendmmsg.bpf.c:
```diff
@@ -38,26 +37,131 @@ int BPF_PROG(sendmmsg_e,
+		return 0;
+	}
+
+	for(int i = 0; i < ret && i < MAX_IOVCNT; i++)
```
should this be normalised to max 12? seems that `MAX_IOVCNT` can be 32

Molter73 replied:
This was also trial and error, logic here used to be a bit smaller and made it possible to do more loops, I'll do some more testing in here, it's possible we can do more loops. I'll also add a comment here.

Molter73 replied:
I managed to up recvmmsg to `MAX_IOVCNT`, so I set both to that value and added a comment.

erthalion commented on userspace/libsinsp/parsers.cpp:
```diff
@@ -4232,7 +4244,13 @@ void sinsp_parser::parse_rw_exit(sinsp_evt *evt)
+				// sendmmsg has all data in the exit event itself.
+				if (etype == PPME_SOCKET_SENDMMSG_X)
+				{
+					enter_evt = evt;
```
I'm curious, why is this a safe thing to do, and why not to retried the enter event? If an enter and exit event will contains the same data that will for sure break some assumptions down the line. An innocent example of what I'm thinking about is the latency calculation, where the timestamps are compared.

Molter73 replied:
Good question! First off, I had to change how this syscall works, usually Falco expects their syscalls to be like:
```
enter
exit
enter
exit
...
``` 
Because I had to loop over messages and send multiple events, this ended up looking like:
```
enter
enter
enter
...
exit
exit
exit
...
```
This broke assumptions made by Falco and resulted in weird behavior like assigning the same IP tuple to all events if the tuple was sent from the enter event. In order to work around this, I had to make sendmmsg submit empty events in the enter event and handle all arguments in the exit event (see [here](https://github.com/stackrox/falcosecurity-libs/blob/985c1c1cf3ba4e1fee92919f845d2238f66a2751/driver/event_table.c#L103-L104)). The recvmmsg is slightly different because it only issues the fd as the arguments for the enter event and this is correct, because that fd is actually used for all messages, so it goes:
```
enter
exit
exit
exit
...
```

Now, why is _this_ line in particular safe? Well, it's only used in two places before returning:
https://github.com/stackrox/falcosecurity-libs/blob/985c1c1cf3ba4e1fee92919f845d2238f66a2751/userspace/libsinsp/parsers.cpp#L4258
If you look at the `update_fd` definition, the second parameter is a const pointer and, looking at the implementation, copies of the required fields are added to the event.

https://github.com/stackrox/falcosecurity-libs/blob/985c1c1cf3ba4e1fee92919f845d2238f66a2751/userspace/libsinsp/parsers.cpp#L4289
Here, `evt->get_fd_info()->m_name` is a `std::string`, so the assignment will make a copy of the 0-th element returned by `enter_evt->get_param_as_str(...)`, so it's also safe.

Any other handling specific to sendmmsg is addressed in this parsing function and some of it in the `sinsp_parser::reset` method ([here](https://github.com/stackrox/falcosecurity-libs/blob/985c1c1cf3ba4e1fee92919f845d2238f66a2751/userspace/libsinsp/parsers.cpp#L509))
https://github.com/stackrox/falcosecurity-libs/blob/985c1c1cf3ba4e1fee92919f845d2238f66a2751/userspace/libsinsp/parsers.cpp#L796-L800

As for the latency calculation... Yeah, that is gonna be broken for this syscall, as stated at the beginning of this mile long comment, the enter/exit event assumption from Falco has to be broken for these syscalls to be tracked. If you have an idea on how we could unbreak it, that'd be great.

erthalion commented:
Overall looks much more straightforward than I expected :) Couple more high-level questions:

* Does this PR contains the Falco bump as well? It seems like at least the bpf verifier log infrastructure was updated, right?
* Looking at the data, gathered by the probe, Collector is interested in everything, except the actual data sent, correct? Would it be possible to optionally avoid copying of the data arguments to reduce the overhead?

Molter73 committed:
Minor cleanup

Molter73 commented:
> Overall looks much more straightforward than I expected :) 

Thanks! (Yes, I'm taking this as a compliment 😄 )

> * Does this PR contains the Falco bump as well? It seems like at least the bpf verifier log infrastructure was updated, right?

The bpf verifier log is added to version 0.18, we are still on 0.17, I had to do a bit of verifier wrestling so I added it in and "forgot" to remove it. In all honesty, I don't think there's any harm in adding it to our changes, they should be dropped on the next bump and are incredibly useful, WDYT? I can get rid of it if you think it can cause problems.

>  * Looking at the data, gathered by the probe, Collector is interested in everything, except the actual data sent, correct? Would it be possible to optionally avoid copying of the data arguments to reduce the overhead?

The data argument is copied up to the `snaplen` which we set to `0` [here](https://github.com/stackrox/collector/blob/60447e570caedaeeed1219fc5377fc3a01ff9257/collector/lib/system-inspector/Service.cpp#L77), so we should be good in that regard.

erthalion commented:
> > Overall looks much more straightforward than I expected :)
> 
> Thanks! (Yes, I'm taking this as a compliment 😄 )

It is, I would produce much more messy implementation :)

> > * Does this PR contains the Falco bump as well? It seems like at least the bpf verifier log infrastructure was updated, right?
> 
> The bpf verifier log is added to version 0.18, we are still on 0.17, I had to do a bit of verifier wrestling so I added it in and "forgot" to remove it. In all honesty, I don't think there's any harm in adding it to our changes, they should be dropped on the next bump and are incredibly useful, WDYT? I can get rid of it if you think it can cause problems.

It makes sense to add it, I'm just in favour of doing it in a separate PR for clarify, if you don't mind.

> > * Looking at the data, gathered by the probe, Collector is interested in everything, except the actual data sent, correct? Would it be possible to optionally avoid copying of the data arguments to reduce the overhead?
> 
> The data argument is copied up to the `snaplen` which we set to `0` [here](https://github.com/stackrox/collector/blob/60447e570caedaeeed1219fc5377fc3a01ff9257/collector/lib/system-inspector/Service.cpp#L77), so we should be good in that regard.

Sounds good!

Molter73 commented:
> It makes sense to add it, I'm just in favour of doing it in a separate PR for clarify, if you don't mind.

Makes sense, I'll get on it!

Molter73 committed:
Rollback verifier logging improvements

Moved the changes to #85.

Stringy reviewed (approved):
LGTM!

Molter73 committed:
Only load interesting syscalls (#84)
